### PR TITLE
*: make `default_authentication_plugin` more compatible with mysql

### LINKED ISSUE #54138: `default_authentication_plugin` is not compatable with MySQL's
Nowadays, [`default_authentication_plugin` in tidb](https://docs.pingcap.com/tidb/stable/system-variables#default_authentication_plugin) only affects the connection between servers and clients. It cannot control the default auth plugin [when creating users without explicit auth plugin as in MySQL](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_default_authentication_plugin). Since `caching_sha2_password` is safer than the default `mysql_native_password`, it is better for tidb to support this compatibility with MySQL.

> It determines which authentication plugin the server assigns to new accounts created by [CREATE USER](https://dev.mysql.com/doc/refman/8.0/en/create-user.html) statements that do not explicitly specify an authentication plugin.

However, since currently using the `caching_sha2_password` in TiDB [requires a SSL connection](https://github.com/pingcap/tidb/blob/4cf86e499364c358fc9420463cccd4ff66240ef6/pkg/server/conn.go#L686), it is not suitable for TiDB to make it default authentication plugin after bootstrapping a new TiDB cluster. It is better to make it default after implementing complete `caching_sha2_password` on TiDB.

### PR DESCRIPTION
**Pull request by CbcWestwolf**
### What problem does this PR solve?

Issue Number: close #54138

Problem Summary:

See the issue decription

### What changed and how does it work?

Get default auth plugin when creating, altering and granting user.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
make `default_authentication_plugin` control default auth plugin when creating user
```

### CODE REVIEW

**Review by lcwangchao — APPROVED** [review]

File: tests/integrationtest/t/executor/simple.test [hunk 1]
```diff
@@ -487,3 +487,30 @@ rollback;
 select * from auto_new;
 
 set autocommit = default;
+
+# TestDefaultAuthPluginForCreateUser
+
+connection default;
+set global default_authentication_plugin = 'caching_sha2_password';
+create user default_sha256_user;
```
Comment (dveeden) [inline tests/integrationtest/t/executor/simple.test:495]: I think the `sha256` is confusing here as there used to be a `sha256_password` plugin.

```suggestion
create user default_sha2_user;
```
(more modifications needed if you merge this)

File: tests/integrationtest/t/executor/simple.test [hunk 2]
```diff
@@ -487,3 +487,30 @@ rollback;
 select * from auto_new;
 
 set autocommit = default;
+
+# TestDefaultAuthPluginForCreateUser
+
+connection default;
+set global default_authentication_plugin = 'caching_sha2_password';
```
Comment (dveeden) [inline tests/integrationtest/t/executor/simple.test:494]: Maybe add something like `set global default_authentication_plugin = 'xxxxx'` to test the validation of the plugin name?

Maybe add statements like `CREATE USER ... IDENTIFIED WITH ...` to test if you can still create users with `mysql_native_plugin` hen the `default_authentication_plugin` is set to `caching_sha2_password`?

File: tests/integrationtest/t/executor/simple.test (hunk 1 above)
Comment (CbcWestwolf) [reply tests/integrationtest/t/executor/simple.test]: updated in https://github.com/pingcap/tidb/commit/fa4a9a166f1a4a2c5b43602dab451f4c87d57ada

File: tests/integrationtest/t/executor/simple.test (hunk 2 above)
Comment (CbcWestwolf) [reply tests/integrationtest/t/executor/simple.test:505]: updated in https://github.com/pingcap/tidb/commit/fa4a9a166f1a4a2c5b43602dab451f4c87d57ada

**Review by dveeden — APPROVED** [review]

**Pull request merged**